### ISSUE TITLE
Conditional operators: .when() / .unless() / .tap()

### ISSUE THREAD

**Issue by ed-fruty** [knex/knex#3818]
# Feature discussion / request 
Very often I saw  conditional cases with knex, like:

```js
const query =knex('users').where({first_name: 'Test'});

if (request.has('company_id')) {
   query.where('company_id', request.get('company_id'));
}

if (! request.has('status')) {
   query.where('status', 'active');
}

...
```

I think it may be simpler and easy to read and support such code without breaking the chain. It may be something like:

```js

knex('users').where({first_name: 'Test'})
  // Conditional execution when the first argument is TRUE
  .when(request.has('company_id'), (query) => query.where(...request.only('company_id')))

  // Conditional execution when the first argument is FALSE
  .unless(request.has('status'), (query) => query.where({status: 'active'}))

  // Always executes, without condition checking, to save the chain.
  .tap((query) => {
      const age  = request.get('age');

      query.when(age > 18, (query) => query.where({ adult : true }));
  })
...

```

Something like that was realized in [Laravel framework](https://laravel.com/docs/7.x/queries#conditional-clauses)

Sorry, if it's duplicates some existing feature request, but I don't see anyone.

**Comment by elhigu** [comment]
Could these be implemented as a separate npm module which can be loaded as a knex plugin. API is found in docs just below https://knexjs.org/#Builder-queryContext

We can list that extension in knex docs, so that people can find it easier. 

Would be great in a long run to be able to make knex core functionality even slimmer and at the same time more robust and better tested so that it will be good platform for others to build on top of it...

**Comment by ed-fruty** [comment]
Yep, but in another hand it can be much simple but powerful to get such feature out of the box, isn't it ?

**Comment by wubzz** [comment]
For all intents and purposes I would opt for `.if(condition, ...args)` / `.ifNot(condition, ...args)` over `.when(condition, ...args)` / `.unless(condition, ...args)`.

**Comment by mercmobily** [comment]
Thanks for the idea. Conditional chaining reads nicely in practice.

This is a feature request with existing alternatives. Today you can get very close ergonomically with `.modify((qb) => condition && qb.where(...))`, with multiple `modify` calls or small helpers for `.when()`/`.unless()` in userland. Adding dedicated `.when()/.unless()/.tap()` sugar to core would be possible, but we try to keep the public surface tight.

Code context: all builders accept callbacks (e.g. `.where((qb) => ...)`) and `.modify` lets you wrap conditional logic without breaking chainability (lib/query/querybuilder.js).

I would actually like to see this in knex... but we will see.
